bin.js

```diff
@@ -0,0 +1,386 @@
+    console.log(`## react-native:copy-build [src=./dist target=${target}]`)
+    execSync(`cp -r ${packageLocation} ${target}`)
+
+    // Start up the local server that will be used to collect logs
+    // Logs will be sent in a single HTTP request
+    console.log('## local-server:start')
+    const { server, close } = await createServerAsync((req, res) => {
+      // Collect test result data
+      const result = []
+      req.on('data', data => result.push(data))
+      req.on('error', (error) => {
+        console.error(error)
+        res.end('fail')
+      })
+      // Once the request data has finished, parse it as JSON and resolve the output
+      req.on('end', () => {
+        res.end('ok')
+        try {
+          const parsed = JSON.parse(Buffer.concat(result).toString('utf8'))
+          _resolve(parsed)
+        } catch (e) {
+          _reject(e)
+        }
+      })
+    }, 1234)
+
+    let driver
+
+    try {
+      // Start an NGROK tunnel for our log collecting server
+      console.log('## ngrok:connect')
+      const publicURL = await ngrok.connect({
+        proto: 'http',
+        region: process.env.NGROK_REGION,
+        addr: server.address().port
+      })
+      console.log(`## ngrok:connected [publicURL=${publicURL}]`)
+
+      // We will generate a JS file which will contain test info
+      // It will export a function to run the package's tests
+      // And it will export the public URL of the NGROK server to send logs to
+      console.log('## react-native:build:prepare')
+      await writeFile(
+      `${root}/test-config.js`,
+      `// This file was generated by rn-tape
+        function runTest () {
+          require('${packageJSON.name}${test}')
+        }
+
+        const publicURL = ${JSON.stringify(publicURL)}
+
+        module.exports = {runTest, publicURL}
+      `
+      )
+
+      let buildDetails
+
+      // If we're running on Android, build the APK for it
+      if (android) {
+        console.log('## react-native:build:android')
+        await logExec('./gradlew', ['assembleRelease'], { cwd: `${root}/android` })
+        buildDetails = {
+          app: `${root}/android/app/build/outputs/apk/release/app-release.apk`,
+          capabilities: {
+            device,
+            os_version: osVersion
+          }
+        }
+      }
+
+      // If we're running on expo, build the XDL bundle
+      if (expo) {
+        console.log('## react-native:build:expo')
+        const build = await parseExpoBuild('npx', ['expo', 'build:android', '-t', 'app-bundle', '--non-interactive'], { cwd: root })
+        console.log({ build })
+        const app = (await exec('node', ['-e', `
+        const xdl = require('@expo/xdl');
+        xdl.Project.getBuildStatusAsync(
+          '.',
+          { platform: 'android', current: false }
+        ).then(
+          data =>
+            console.log(
+              data.jobs?.filter(job => job.id === '${build}')[0]?.artifacts.url
```

Comment:
This requires node v14. Which is okay, but an error message a little earlier would be good.

Reply:
Oh snap, I didn't even notice. I might refactor to avoid using optional chaining if that's okay.

Reply:
Sure, both good ways to go!